triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala

```diff
@@ -132,10 +138,34 @@ object Trigger extends StrictLogging {
       }
     } yield hasReadAs
 
-  def fromIdentifier(
-      compiledPackages: CompiledPackages,
+  private[trigger] def detectVersion(
+      pkgInterface: PackageInterface,
+      triggerIds: TriggerIds,
+  ): Either[String, Trigger.Version] = {
+    import scalaz.std.either._
+    import scalaz.syntax.traverse._
+
+    val versionId = triggerIds.damlTriggerInternal("version")
+    val versionType = pkgInterface.lookupValue(versionId).toOption.map(_.typ)
+    for {
+      versionStr <- versionType.traverseU {
+        case TTyCon(versionTypCon @ Identifier(_, QualifiedName(_, name)))
+            if versionTypCon == triggerIds.damlTriggerInternal(name.segments.head) =>
+          Right(name.segments.head)
+        case _ =>
+          Left(s"cannot infer trigger version")
+      }
+      version <- Trigger.Version.fromString(versionStr)
+    } yield version
+  }
```

Comment:
Perhaps the following would be simpler?

```suggestion
  private[trigger] def detectVersion(
      pkgInterface: PackageInterface,
      triggerIds: TriggerIds,
  ): Either[String, Trigger.Version] = {
    val versionId = triggerIds.damlTriggerInternal("version")
    val versionType = pkgInterface.lookupValue(versionId).toOption.map(_.typ)
    for {
      versionStr <- versionType.fold[Either[String, Option[Name]]](Right(None)) {
        case TTyCon(versionTypCon @ Identifier(_, QualifiedName(_, name)))
          if versionTypCon == triggerIds.damlTriggerInternal(name.segments.head) =>
          Right(Some(name.segments.head))
        case _ =>
          Left("can not infer trigger version")
      }
      version <- Trigger.Version.fromString(versionStr)
    } yield version
  }
```